main/src/ca/uwaterloo/flix/language/phase/Parser.scala

```diff
@@ -838,17 +838,37 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       }
     }
 
-    def TryCatch: Rule1[ParsedAst.Expression] = {
+    def Do: Rule1[ParsedAst.Expression] = rule {
+      SP ~ keyword("do") ~ WS ~ Names.QualifiedEffect ~ ArgumentList ~ SP ~> ParsedAst.Expression.Do
+    }
+
+    def Resume: Rule1[ParsedAst.Expression] = rule {
+      SP ~ keyword("resume") ~ ArgumentList ~ SP ~> ParsedAst.Expression.Resume
+    }
+
+    def Try: Rule1[ParsedAst.Expression] = {
       def CatchRule: Rule1[ParsedAst.CatchRule] = rule {
         keyword("case") ~ WS ~ Names.Variable ~ optWS ~ ":" ~ optWS ~ atomic("##") ~ Names.JavaName ~ WS ~ atomic("=>") ~ optWS ~ Expression ~> ParsedAst.CatchRule
       }
 
-      def CatchBody: Rule1[Seq[ParsedAst.CatchRule]] = rule {
-        "{" ~ optWS ~ oneOrMore(CatchRule).separatedBy(CaseSeparator) ~ optWS ~ "}"
+      def CatchBody: Rule1[ParsedAst.Expression.CatchOrHandler] = rule {
+        keyword("catch") ~ optWS ~ "{" ~ optWS ~ oneOrMore(CatchRule).separatedBy(CaseSeparator) ~ optWS ~ "}" ~> ParsedAst.Expression.CatchOrHandler.Catch
+      }
+
+      def HandlerRule: Rule1[ParsedAst.HandlerRule] = rule {
+        keyword("def") ~ WS ~ Names.QualifiedEffect ~ FormalParamList ~ optWS ~ atomic("=") ~ optWS ~ Expression ~> ParsedAst.HandlerRule
+      }
+
+      def HandlerBody: Rule1[ParsedAst.Expression.CatchOrHandler] = rule {
+        keyword("with") ~ optWS ~ "{" ~ optWS ~ oneOrMore(HandlerRule).separatedBy(CaseSeparator) ~ optWS ~ "}" ~> ParsedAst.Expression.CatchOrHandler.Handler
```

Comment:
Can an effect have zero operations? I guess not hence why you require at least one handler? But still, why not allow zero handlers? It should be detected later.